tests/integration/rdb.tcl

```diff
@@ -416,4 +416,157 @@ start_server {} {
+            wait_done_loading r
+
+            assert_equal [r hgetall key] {}
+        }
+    }
+}
+
+set server_path [tmpdir "server.all-hfield-exp-test"]
+
+# verifies a long TTL value (6 bytes) is saved and loaded correctly
+start_server [list overrides [list "dir" $server_path]] {
+    foreach type {listpack dict} {
+        test "hash field expiration save and load rdb long TTL, ($type)" {
+            if {$type eq "dict"} {
+                r config set hash-max-listpack-entries 0
+            } else {
+                r config set hash-max-listpack-entries 512
+            }
+
+            r FLUSHALL
+
+            r HSET key a 1
+            # set expiry to 0xabcdef987654 (6 bytes)
+            r HPEXPIREAT key 188900976391764 1 a
+
+            r save
+            restart_server 0 true false
+            wait_done_loading r
+
+            assert_equal [r hget key a ] 1
+            assert_equal [r hpexpiretime key 1 a] {188900976391764}
+        }
+    }
+}
+
+set server_path [tmpdir "server.listpack-to-dict-test"]
+
+test "save listpack, load dict" {
+    start_server [list overrides [list "dir" $server_path  enable-debug-command yes]] {
+        r config set hash-max-listpack-entries 512
+
+        r FLUSHALL
+
+        r HMSET key a 1 b 2 c 3 d 4
+        assert_match "*encoding:listpack*" [r debug object key]
+        r HPEXPIRE key 100 1 d
+
+        # sleep 100 ms to make sure all fields will expire after restart
+        after 100
+
+        # change configuration and reload - result should be dict-encoded key
+        r config set hash-max-listpack-entries 0
+        r debug reload
+
+        assert_equal [r hget key a] 1
+        assert_equal [r hget key b] 2
+        assert_equal [r hget key c] 3
+        assert_equal [r hget key d] {}
+        assert_match "*encoding:hashtable*" [r debug object key]
+    }
+}
+
+set server_path [tmpdir "server.dict-to-listpack-test"]
+
+# TODO: re-visit when removing active expiration on load for listpack
+test "save dict, load listpack" {
+    start_server [list overrides [list "dir" $server_path  enable-debug-command yes]] {
+        r config set hash-max-listpack-entries 0
+
+        r FLUSHALL
+
+        r HMSET key a 1 b 2 c 3 d 4
+        assert_match "*encoding:hashtable*" [r debug object key]
+        r HPEXPIRE key 100 1 d
+
+        # sleep 100 ms to make sure all fields will expire after restart
+        after 100
+
+        # change configuration and reload - result should be LP-encoded key
+        r config set hash-max-listpack-entries 512
+        r debug reload
+
+        assert_equal [r hget key a] 1
+        assert_equal [r hget key b] 2
+        assert_equal [r hget key c] 3
+        assert_equal [r hget key d] {}
+        assert_match "*encoding:listpack*" [r debug object key]
+    }
+}
+
+
+#Also:
+#2. try listpack verification during load
+#3. think about listpack with duplicated fields when one of them is already expired: decision: Need to fail! verify
+
```

Comment:
We need a finer resolution of expiry after restart, whether the item was expired:
1.  During RDB load
2.  Shortly after by active-expiration 
3. Later by lazy expiration.

And apply this test for listpack and dict.

Reply:
This is anyway problematic considering how tests are written.
When you "hget" or "hgetall", active expiry spring into action, doesn't it? So it's not really checking expiry on load...
Is there a way to disable it?